### ISSUE TITLE
[Bug]: Controlnet slow hooked time

### ISSUE THREAD

**Issue by mchong6** [Mikubill/sd-webui-controlnet#2243]
### Is there an existing issue for this?

- [X] I have searched the existing issues and checked the recent builds/commits of both this extension and the webui

### What happened?

I am using webui api to batch generate images. I am using multiple controlnets, ip adapter face plus and openpose. However, between each generation, the controlnet hooked time is really slow. I don't seem to recall it being this slow in the past. I tried playing with control_net_model_cache_size and control_net_unit_count but those do not help.

### Steps to reproduce the problem

1. Run webui api 
2. use txt2img and run ipadaptor and controlnet openpose multiple times

### What should have happened?

controlnet hooked time should be lower

### Commit where the problem happens

webui: version: v1.6.0-2-g4afaaf8a  •  python: 3.9.2  •  torch: 2.0.1+cu118  •  xformers: N/A  •  gradio: 3.41.2  •  checkpoint: caa95002a1
controlnet: [05ef0b1c](https://github.com/Mikubill/sd-webui-controlnet/commit/05ef0b1cd1374cf285dd8d5ccd7db9997549893c)

### What browsers do you use to access the UI ?

Google Chrome

### Command Line Arguments

```Shell
--opt-sdp-attention --api
```

### List of enabled extensions

<img width="867" alt="Screenshot 2023-11-07 at 2 23 06 AM" src="https://github.com/Mikubill/sd-webui-controlnet/assets/20831135/261714a1-fe80-4d85-bad2-e10cdc21dd17">

### Console logs

```Shell
Reusing loaded model icbinpICantBelieveIts_secoInpainting.safetensors [c1eb2d6144] to load epicrealism_pureEvolutionV5-inpainting.safetensors [caa95002a1
]                                                                                                                                                        
Loading weights [caa95002a1] from /opt/tiger/stable-diffusion-webui/models/Stable-diffusion/epicrealism_pureEvolutionV5-inpainting.safetensors           
Applying attention optimization: sdp... done.                                                                                                            
Weights loaded in 1.5s (send model to cpu: 0.9s, load weights from disk: 0.2s, apply weights to model: 0.1s, move model to device: 0.3s).                
2023-11-07 17:55:13,100 - ControlNet - INFO - Loading model: ip-adapter-plus-face_sd15 [71693645]                                                        
2023-11-07 17:55:13,130 - ControlNet - INFO - Loaded state_dict from [/opt/tiger/stable-diffusion-webui/models/ControlNet/ip-adapter-plus-face_sd15.pth] 
2023-11-07 17:55:13,375 - ControlNet - INFO - ControlNet model ip-adapter-plus-face_sd15 [71693645] loaded.                                              
2023-11-07 17:55:13,387 - ControlNet - DEBUG - Safe numpy convertion START                                                                               
2023-11-07 17:55:13,387 - ControlNet - DEBUG - Safe numpy convertion END                                                                                 
2023-11-07 17:55:13,387 - ControlNet - INFO - Loading preprocessor: ip-adapter_clip_sd15                                                                 
2023-11-07 17:55:13,387 - ControlNet - INFO - preprocessor resolution = 512                                                                              
2023-11-07 17:55:13,388 - ControlNet - DEBUG - Use numpy seed 1896900464.                                                                                
2023-11-07 17:55:13,388 - ControlNet - DEBUG - Calling preprocessor ip-adapter_clip_sd15 outside of cache.                                               
2023-11-07 17:55:21,750 - ControlNet - INFO - Loading model: control_v11p_sd15_openpose [cab727d4]                                                       
2023-11-07 17:55:21,983 - ControlNet - INFO - Loaded state_dict from [/opt/tiger/stable-diffusion-webui/extensions/sd-webui-controlnet/models/control_v11
p_sd15_openpose.pth]                                                                                                                                     
2023-11-07 17:55:21,983 - ControlNet - INFO - controlnet_default_config                                                                                  
2023-11-07 17:55:23,924 - ControlNet - INFO - ControlNet model control_v11p_sd15_openpose [cab727d4] loaded.                                             
2023-11-07 17:55:23,937 - ControlNet - DEBUG - Safe numpy convertion START                                                                               
2023-11-07 17:55:23,937 - ControlNet - DEBUG - Safe numpy convertion END                                                                                 
2023-11-07 17:55:23,937 - ControlNet - INFO - Loading preprocessor: openpose_full                                                                        
2023-11-07 17:55:23,937 - ControlNet - INFO - preprocessor resolution = 512                                                                              
2023-11-07 17:55:23,937 - ControlNet - DEBUG - Use numpy seed 1896900464.                                                                                
2023-11-07 17:55:23,938 - ControlNet - DEBUG - Calling preprocessor openpose_full outside of cache.                                                      
2023-11-07 17:55:26,153 - ControlNet - DEBUG - controlnet_main_entry ran in: 13.288 sec                                                                  
2023-11-07 17:55:26,153 - ControlNet - INFO - ControlNet Hooked - Time = 13.28781247138977                                                               
2023-11-07 17:55:26,153 - ControlNet - DEBUG - controlnet_hack ran in: 13.288 sec                                                                        
2023-11-07 17:55:26,153 - ControlNet - DEBUG - process ran in: 13.288 sec                                                                                
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:03<00:00,  6.40it/s]
2023-11-07 17:55:30,943 - ControlNet - DEBUG - postprocess ran in: 0.268 sec█████████████████████████████████████████████| 25/25 [00:03<00:00,  7.67it/s]
Total progress: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:03<00:00,  7.26it/s]
2023-11-07 17:55:32,111 - ControlNet - INFO - Loading model from cache: ip-adapter-plus-face_sd15 [71693645]█████████████| 25/25 [00:03<00:00,  7.67it/s]
2023-11-07 17:55:32,123 - ControlNet - DEBUG - Safe numpy convertion START                                                                               
2023-11-07 17:55:32,123 - ControlNet - DEBUG - Safe numpy convertion END                                                                                 
2023-11-07 17:55:32,123 - ControlNet - INFO - Loading preprocessor: ip-adapter_clip_sd15                                                                 
2023-11-07 17:55:32,123 - ControlNet - INFO - preprocessor resolution = 512                                                                              
2023-11-07 17:55:32,123 - ControlNet - DEBUG - Use numpy seed 3274869290.
2023-11-07 17:55:32,124 - ControlNet - DEBUG - Calling preprocessor ip-adapter_clip_sd15 outside of cache.
2023-11-07 17:55:40,935 - ControlNet - INFO - Loading model from cache: control_v11p_sd15_openpose [cab727d4]
2023-11-07 17:55:40,947 - ControlNet - DEBUG - Safe numpy convertion START
2023-11-07 17:55:40,947 - ControlNet - DEBUG - Safe numpy convertion END
2023-11-07 17:55:40,947 - ControlNet - INFO - Loading preprocessor: openpose_full
2023-11-07 17:55:40,947 - ControlNet - INFO - preprocessor resolution = 512
2023-11-07 17:55:40,948 - ControlNet - DEBUG - Use numpy seed 3274869290.
2023-11-07 17:55:40,948 - ControlNet - DEBUG - Calling preprocessor openpose_full outside of cache.
2023-11-07 17:55:42,221 - ControlNet - DEBUG - controlnet_main_entry ran in: 10.199 sec
2023-11-07 17:55:42,222 - ControlNet - INFO - ControlNet Hooked - Time = 10.199609279632568
2023-11-07 17:55:42,222 - ControlNet - DEBUG - controlnet_hack ran in: 10.200 sec
2023-11-07 17:55:42,222 - ControlNet - DEBUG - process ran in: 10.200 sec
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:03<00:00,  7.68it/s]
2023-11-07 17:55:46,149 - ControlNet - DEBUG - postprocess ran in: 0.272 sec█████████████████████████████████████████████| 25/25 [00:03<00:00,  7.78it/s]
Total progress: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:03<00:00,  7.19it/s]
2023-11-07 17:55:47,080 - ControlNet - INFO - Loading model from cache: ip-adapter-plus-face_sd15 [71693645]█████████████| 25/25 [00:03<00:00,  7.78it/s]
2023-11-07 17:55:47,091 - ControlNet - DEBUG - Safe numpy convertion START
2023-11-07 17:55:47,092 - ControlNet - DEBUG - Safe numpy convertion END
2023-11-07 17:55:47,092 - ControlNet - INFO - Loading preprocessor: ip-adapter_clip_sd15
2023-11-07 17:55:47,092 - ControlNet - INFO - preprocessor resolution = 512
2023-11-07 17:55:47,092 - ControlNet - DEBUG - Use numpy seed 4183745466.
2023-11-07 17:55:47,092 - ControlNet - DEBUG - Calling preprocessor ip-adapter_clip_sd15 outside of cache.
2023-11-07 17:55:55,004 - ControlNet - INFO - Loading model from cache: control_v11p_sd15_openpose [cab727d4]
2023-11-07 17:55:55,015 - ControlNet - DEBUG - Safe numpy convertion START
2023-11-07 17:55:55,015 - ControlNet - DEBUG - Safe numpy convertion END
2023-11-07 17:55:55,015 - ControlNet - INFO - Loading preprocessor: openpose_full
2023-11-07 17:55:55,015 - ControlNet - INFO - preprocessor resolution = 512
2023-11-07 17:55:55,015 - ControlNet - DEBUG - Use numpy seed 4183745466.
2023-11-07 17:55:55,016 - ControlNet - DEBUG - Calling preprocessor openpose_full outside of cache.
2023-11-07 17:55:56,270 - ControlNet - DEBUG - controlnet_main_entry ran in: 9.257 sec
2023-11-07 17:55:56,270 - ControlNet - INFO - ControlNet Hooked - Time = 9.257014989852905
2023-11-07 17:55:56,270 - ControlNet - DEBUG - controlnet_hack ran in: 9.257 sec
2023-11-07 17:55:56,270 - ControlNet - DEBUG - process ran in: 9.257 sec
```

### Additional information
My controlnet settings
<img width="521" alt="Screenshot 2023-11-07 at 2 28 03 AM" src="https://github.com/Mikubill/sd-webui-controlnet/assets/20831135/d824490b-d211-411c-8d03-abd2f0e46c20">

**Comment by huchenlei** [comment]
From the log it seems like IP Adapter preprocessor is taking the majority of time.

Can you try having only the IP Adapter unit enabled and see if hook time shortens?

Note:
After initial generation, the preprocessor result is cached, subsequent generations should be faster as long as IP Adapter input is not changed. The impression you had that ControlNet runs fast might due to caching.

**Comment by mchong6** [comment]
Using only IP adaptor is still slow 

> Weights loaded in 1.7s (send model to cpu: 1.1s, load weights from disk: 0.2s, apply weights to model: 0.1s, move model to device: 0.3s).
2023-11-08 10:59:14,113 - ControlNet - INFO - Loading model: ip-adapter-plus-face_sd15 [71693645]
2023-11-08 10:59:14,143 - ControlNet - INFO - Loaded state_dict from [/opt/tiger/stable-diffusion-webui/models/ControlNet/ip-adapter-plus-face_sd15.pth]
2023-11-08 10:59:14,396 - ControlNet - INFO - ControlNet model ip-adapter-plus-face_sd15 [71693645] loaded.
2023-11-08 10:59:14,410 - ControlNet - DEBUG - Safe numpy convertion START
2023-11-08 10:59:14,410 - ControlNet - DEBUG - Safe numpy convertion END
2023-11-08 10:59:14,410 - ControlNet - INFO - Loading preprocessor: ip-adapter_clip_sd15
2023-11-08 10:59:14,410 - ControlNet - INFO - preprocessor resolution = 512
2023-11-08 10:59:14,410 - ControlNet - DEBUG - Use numpy seed 1354590066.
2023-11-08 10:59:24,179 - ControlNet - DEBUG - controlnet_hack ran in: 10.331 sec
2023-11-08 10:59:24,179 - ControlNet - DEBUG - process ran in: 10.331 sec
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:02<00:00,  8.55it/s]
2023-11-08 10:59:28,052 - ControlNet - DEBUG - postprocess ran in: 0.297 sec█████████████████████████████████████████████| 25/25 [00:02<00:00, 10.87it/s]
Total progress: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:02<00:00,  9.87it/s]
2023-11-08 10:59:29,077 - ControlNet - INFO - Loading model from cache: ip-adapter-plus-face_sd15 [71693645]█████████████| 25/25 [00:02<00:00, 10.87it/s]
2023-11-08 10:59:29,090 - ControlNet - DEBUG - Safe numpy convertion START
2023-11-08 10:59:29,090 - ControlNet - DEBUG - Safe numpy convertion END
2023-11-08 10:59:29,090 - ControlNet - INFO - Loading preprocessor: ip-adapter_clip_sd15
2023-11-08 10:59:29,090 - ControlNet - INFO - preprocessor resolution = 512
2023-11-08 10:59:29,090 - ControlNet - DEBUG - Use numpy seed 1036719364.
2023-11-08 10:59:29,091 - ControlNet - DEBUG - Calling preprocessor ip-adapter_clip_sd15 outside of cache.
2023-11-08 10:59:38,085 - ControlNet - DEBUG - controlnet_main_entry ran in: 9.016 sec
2023-11-08 10:59:38,086 - ControlNet - INFO - ControlNet Hooked - Time = 9.016075372695923
2023-11-08 10:59:38,086 - ControlNet - DEBUG - controlnet_hack ran in: 9.016 sec
2023-11-08 10:59:38,086 - ControlNet - DEBUG - process ran in: 9.016 sec
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:02<00:00, 11.17it/s]
2023-11-08 10:59:41,007 - ControlNet - DEBUG - postprocess ran in: 0.275 sec█████████████████████████████████████████████| 25/25 [00:02<00:00, 10.80it/s]
Total progress: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:02<00:00,  9.96it/s]
2023-11-08 10:59:41,881 - ControlNet - INFO - Loading model from cache: ip-adapter-plus-face_sd15 [71693645]█████████████| 25/25 [00:02<00:00, 10.80it/s]
2023-11-08 10:59:41,891 - ControlNet - DEBUG - Safe numpy convertion START
2023-11-08 10:59:41,892 - ControlNet - DEBUG - Safe numpy convertion END
2023-11-08 10:59:41,892 - ControlNet - INFO - Loading preprocessor: ip-adapter_clip_sd15
2023-11-08 10:59:41,892 - ControlNet - INFO - preprocessor resolution = 512
2023-11-08 10:59:41,892 - ControlNet - DEBUG - Use numpy seed 2726189318.
2023-11-08 10:59:41,892 - ControlNet - DEBUG - Calling preprocessor ip-adapter_clip_sd15 outside of cache.
2023-11-08 10:59:50,341 - ControlNet - DEBUG - controlnet_main_entry ran in: 8.468 sec
2023-11-08 10:59:50,341 - ControlNet - INFO - ControlNet Hooked - Time = 8.468431234359741
2023-11-08 10:59:50,341 - ControlNet - DEBUG - controlnet_hack ran in: 8.468 sec
2023-11-08 10:59:50,341 - ControlNet - DEBUG - process ran in: 8.469 sec

Each generation I have a different IP adaptor input so it makes sense if it wouldn't cache the results. What I am confused about is that it took only 3s hooked time previously when I used the same 2 IP adaptor units and had similar workflow.

**Comment by huchenlei** [comment]
Can you reproduce your previous experience? This might require checkout an older version of the extension. If you are not sure which version you was using, you can follow `git bisect` guide here: https://www.metaltoad.com/blog/beginners-guide-git-bisect-process-elimination

**Comment by mchong6** [comment]
Unfortunately, I couldn't access the old code so I was trying to replicate it now using the same workflow. I remember having a similar problem previously but it got "fixed" after I increased the control_net_model_cache_size and control_net_unit_count.  This time it did not work. So I am guessing the previous slowdown is caused by not loading the model from cache and not the preprocessor.

One surprising thing was that there a few occasions in the batch generation where the hooked time gets down to 0.001s or so. And if I look at the logs, this happens when "Calling preprocessor ip-adapter_clip_sd15 outside of cache" does not happen. Each generation uses a different IP adapter input so I have no idea why this happens. This happened only a few times and the vast majority of time its super slow.

Is there any way I can look into the preprocessor and see whats causing the slowdown?

**Comment by huchenlei** [comment]
A1111 requires each extension to be installed as a git repo so you definitely have the access to older version of the code. You can use ChatGPT to help you sort out all git commands necessary to perform the bisect if you are not familiar with git.

**Comment by TheSlayonetta** [comment]
This is still happening for me. Is there just a way to make this weight system optional since it clearly doesn't work for a significant amount of people? Like for those that it works for great they can still use it but to make it mandatory for everyone and for those of us it doesn't work for don't have to go on some scavenger hunt to find a workaround when we aren't all efficient at coding and have to redo it every time automatic1111 updates?

**Comment by fu-jianhua** [comment]
![12](https://github.com/user-attachments/assets/5b4c6caa-6eee-44f3-8a80-d934e65e3202)
I have also encountered this problem, but it will disappear after restarting. If the service keeps starting for a long time, this problem will appear. I joined ADetailer facial restoration, so I loaded the Controlnet model twice, and both loads took a long time, up to one minute. My Controlnet version is v1.1.441,The computer graphics card driver is Driver Version: 520.61.05 and CUDA Version: 11.8. May I ask if this issue is caused by my driver being too low?
